[multistage] Initial planner changes to support the v1 aggregation functions in multistage

somandal opened the pull request:
This work is a collaboration between @vvivekiyer and myself
*Important Note:* This PR must be rebased on top of @vvivekiyer 's runtime changes (PR: https://github.com/apache/pinot/pull/10845) and integrated before it can be merged
This PR should also be reviewed with @vvivekiyer 's https://github.com/apache/pinot/pull/10845

Motivation:

- Today only a subset of aggregation functions (sum, min, max, count, avg, count(distinct), bool_or, bool_and, kurtosis, and skewness) work in the multistage engine.
- To support each new aggregation function changes are required on both the planner and runtime side. Pinot contributors should not need to understand the nitty gritty details of the planner and execution engine to add new aggregation functions.
- The multistage `AggregateOperator` essentially reimplements the aggregation logic rather than reusing the aggregation logic present in the v1 engine functions. This results in code duplication and a high maintenance overhead (e.g. bug fixes must be made in two places which may even have different logic or different mechanisms to support things like null handling).
- The v1 execution engine supports a large number of aggregation functions and it is not practical to manually add support for each new aggregation function to multistage. This also results in a situation in which each new aggregation function needs to be added to both the v1 engine and the multistage engine, or if support is only added for the v1 engine, the multistage engine must constantly play catch up.

To get around the above limitations we are refactoring the aggregation function framework to work for both the v1 and multistage engine. This PR tackles the changes required on the planner side to make the planner more generic. This is phase 1 of the planned set of changed to support the aggregation functions and tackles only the existing set of aggregation functions already supported in the multistage engine.

The high-level approach is to split the aggregations into 3 LogicalAggregate nodes instead of 2, where each node performs the following operations using the v1 `AggregationFunction` interface methods:

- Leaf Stage LogicalAggregate: Performs `aggregate()` / `aggregateGroupBySV()` / `aggregateGroupByMV()`
    - Input type: original row type. Output type: `getIntermediateResultColumnType()` (which may be an OBJECT)
- Intermediate Stage LogicalAggregate: Performs `merge()`
    - Input type: `getIntermediateResultColumnType()` (which may be an OBJECT, except when skip leaf level hints are used in which case the input type is row type). Output type: `getIntermediateResultColumnType()` (which may be an OBJECT)
- Final Stage LogicalAggregate: Performs `extractFinalResult()` with final type conversion
    - Input type: `getIntermediateResultColumnType()` (which may be an OBJECT). Output type: `getFinalResultColumnType()`

For scenarios where hints are used to skip the leaf level aggregation, the intermediate stage will perform the `aggregate()` / `aggregateGroupBySV()` / `aggregateGroupByMV()` instead.

More details can be found here:

- OSS Issue: https://github.com/apache/pinot/issues/10745
- Design document: https://docs.google.com/document/d/1Us6aBvTpNLMEy0ODo34OgTk73h_LVFFAH6q17689h1M/edit?usp=sharing
- @vvivekiyer 's runtime changes to support this: https://github.com/apache/pinot/pull/10845

Types of functions this PR does *not* support (and for which support will be added later on):

- Add support for taking function arguments as literals (e.g. PERCENTILE)
- Add support for aggregation functions which take more than one column as arguments (e.g . COVARIANCE)
- Arg min/max type aggregation functions using the Parent-Child relationship

Note: This series of changes does not plan to address the MV aggregation functions as we first need to add MV support to multistage, and the aggregations will be tackled as part of that effort. Please see https://github.com/apache/pinot/issues/10658 for more details.

### Steps to merge this Planner PR (after this [Runtime PR](https://github.com/apache/pinot/pull/10845) is merged):

- Rebase and resolve conflicts (this will pull in merged runtime changes)
- Delete the existing AggregateOperator.java. Rename NewAggregateOperator.java to AggregateOperator.java (see the commented code in PhysicalPlanVisitor.visitAggregate())
- Look for “TODO(Sonam)” comments across the codebase (called out files below) and address them:

1. AggregationFunctionFactory.java
2. NewAggregatorOperator.java
3. PhysicalPlanVisitor.java
4. AggregateOperatorTest.java
5. Aggregates.json

- Review, run tests and merge code.

### Future Work:

- Today SUM and SUM0 are handled similarly. In the future, we can have separate v1 aggregation functions SUM and SUM0 where the return types are differentiated. Today Calcite rewrites all SUM as SUM0, and if the types are non-nullable then they do not even add a CASE operator to return null for SUM = 0 (which is true for all scenarios we tested based on the planner tests)
- Today, when the AggregationOperator (in leaf stage) executes aggregations like MIN, MAX, SUM on a column where null handling is disabled, it would explicitly return infinity, -infinity and 0 respectively. However, if the same aggregation is handled in the AggregateOperator (in intermediate stage) we would get null as a result. This discrepancy exists today even without our changes.
- Calcite divides AVG aggregation functions vertically into SUM and COUNT. The division operator (/) is then used to perform SUM / COUNT. However, our v1 aggregation code could return 0 as the default value for COUNT and this will result in division by zero error. This could not be handled without a scalar function today because all our aggregation fields are set to non-nullable types according to Calcite. Using a Calcite CASE operator to return null for COUNT=0 and only then perform division results in a planning error because of type mismatch between nullable vs the expected non-nullable type.
- Need to change some of the functions such as SUM, MIN, MAX to use the appropriate types rather than forcing them to return a double result.
- Need to add support to take literal arguments for the aggregation functions (e.g. PERCENTILE, DISTINCTCOUNTHLL, etc) and add tests for all such functions to ensure they work as expected.
- Add support for certain aggregation functions which take more than one column as argument (e.g. COVARIANCE)
- ArgMin/ArgMax support

cc @siddharthteotia @vvivekiyer @walterddr @kishoreg

somandal commented:
Just a note: the test failures are currently expected as the changes need to be integrated with the runtime code

somandal commented on pinot-query-planner/src/main/java/org/apache/calcite/sql/fun/PinotOperatorTable.java:
```diff
@@ -103,6 +98,61 @@ public final void initNoDuplicate() {
+    // aren't standard Calcite functions such as SUM / MIN / MAX / COUNT etc.
+    for (AggregationFunctionType aggregationFunctionType : AggregationFunctionType.values()) {
+      if (aggregationFunctionType.isNativeCalciteAggregationFunctionType()
+          || !aggregationFunctionType.isSupportedInMultiStage()) {
```
Are you talking about `isSupportedInMultiStage()`? I can perhaps rename it to something like `registerWithOperatorTable()` or something like that for now.

This method is meant to be a temporary one which we will remove once support for all aggregation functions is added to multistage. I don't want to keep a hard-coded list here because eventually most aggregation functions will be supported and the full list is available as part of the `AggregationFunctionType` enum class anyways.

Also note that there are many MV aggregation functions which we cannot support yet without generically adding MV support to multi-stage as part of this OSS issue: https://github.com/apache/pinot/issues/10658 
So we may need to gatekeep which aggregation functions are supported for a while, but over time it should be possible to support most.

I'm not sure what you mean by backward incompatibility when we make v2 the default? Can you elaborate a bit on that part?

kishoreg replied:
then have a negation list to list the ones that are not supported in v2 or something calcite already supports.. my concern is that the aggregationfunction type itself should not care about its supported in multistage etc..

somandal commented:
Have addressed the suggestions from @walterddr and @Jackie-Jiang 

This PR is ready to be reviewed @walterddr @Jackie-Jiang @kishoreg @xiangfu0 but must only be merged after rebasing and incorporating changes from https://github.com/apache/pinot/pull/10845

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/BaseSingleInputAggregationFunction.java:
```diff
@@ -40,7 +40,7 @@ public BaseSingleInputAggregationFunction(ExpressionContext expression) {
   @Override
   public String getResultColumnName() {
-    return getType().getName().toLowerCase() + "(" + _expression + ")";
+    return getType().getName().replace("_", "").toLowerCase() + "(" + _expression + ")";
```
will this be backward incompatible when upgrading 
(e.g. server returns a different result ColunmName while broker is still using older version of the name and thus can't find the right column?)

somandal replied:
none of the v1 aggregation function have an '_' in their name so this should be fine. This was added to handle backward incompatibility for BOOL_AND and BOOL_OR which are named with underscore in v2 but don't have an underscore in v1.

somandal replied:
discussed offline - now we register both names without and without the underscore. removed this change

walterddr commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintStrategyTable.java:
```diff
@@ -27,8 +27,11 @@
  */
 public class PinotHintStrategyTable {
-  public static final String INTERNAL_AGG_INTERMEDIATE_STAGE = "aggIntermediateStage";
   public static final String INTERNAL_AGG_FINAL_STAGE = "aggFinalStage";
+  public static final String INTERNAL_AGG_INTERMEDIATE_STAGE = "aggIntermediateStage";
+  public static final String INTERNAL_AGG_LEAF_STAGE = "aggLeafStage";
```
1. can we add javadocs for these? i think some definition is different from previous usage of these internal hints
2. please also include the internal hints to agg node function name mapping in the java doc and reference to those ( i believe it was called `MODE` in #10845)

somandal replied:
done

walterddr commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateReduceFunctionsRule.java:
```diff
@@ -0,0 +1,427 @@
+ *
+ * @see CoreRules#AGGREGATE_REDUCE_FUNCTIONS
+ */
+public class PinotAggregateReduceFunctionsRule
```
why the rename? is most code similar to PinotReduceAggregateFunctionRule?

somandal replied:
discussed this already and decided not to rename it back as it was added as a replacement for an existing calcite rule and not a replacement for the other one

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/query/reduce/function/InternalReduceFunctions.java:
```diff
@@ -35,17 +33,10 @@ private InternalReduceFunctions() {
-
-  @ScalarFunction
-  public static int countDistinctReduce(Set<?> values) {
-    return values.size();
+  public static Double avgReduce(Double intermediateResultSum, Long intermediateResultCount) {
```
`@ScalarFunction(nullableParameters = true)`

somandal replied:
done

walterddr commented on pinot-query-planner/src/main/java/org/apache/pinot/query/planner/plannode/AggregateNode.java:
```diff
@@ -35,30 +35,72 @@ public class AggregateNode extends AbstractPlanNode {
+  @ProtoProperties
+  private boolean _treatIntermediateStageAsLeaf;
+
+  /**
+   * Enum to denote the aggregation stage being performed. Hints are used to populate these values
+   * LEAF - leaf aggregation level which performs the aggregations on the raw values directly
+   * INTERMEDIATE - intermediate aggregation level which merges results from lower aggregation levels
+   * FINAL - final aggregation level which extracts the final result
+   */
+  public enum AggregationStage {
+    LEAF,
+    INTERMEDIATE,
+    FINAL
+  }
```
let's merge this ENUM with AggregateOperator.Mode
IMO, there's nothing we can't determine during plan time for runtime operator to perform.

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java:
```diff
@@ -100,17 +100,14 @@ public MultiStageOperator visitAggregate(AggregateNode node, PhysicalPlanContext
-//        treatIntermediateAsLeaf);
-    return new AggregateOperator(context.getOpChainExecutionContext(), nextOperator, node.getDataSchema(),
-        node.getAggCalls(), node.getGroupSet(), node.getInputs().get(0).getDataSchema());
+    boolean extractFinalResult = AggregateNode.isFinalStage(node);
+    boolean isIntermediateStage = AggregateNode.isIntermediateStage(node);
+    boolean isLeafStage = AggregateNode.isLeafStage(node);
+    boolean treatIntermediateAsLeaf = node.isTreatIntermediateStageAsLeaf();
```
directly set mode is much simplier see previous comments

walterddr commented on pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java:
```diff
@@ -79,7 +77,8 @@ public void shouldHandleUpstreamErrorBlocks() {
     DataSchema outSchema = new DataSchema(new String[]{"sum"}, new ColumnDataType[]{DOUBLE});
     AggregateOperator operator =
-        new AggregateOperator(OperatorTestUtil.getDefaultContext(), _input, outSchema, calls, group, inSchema);
+        new AggregateOperator(OperatorTestUtil.getDefaultContext(), _input, outSchema, inSchema, calls, group,
+            false, true, false, false);
```
TODO: many of these functions are setup to run in both leaf and intermediate, we need to mix and match test modes as well

xiangfu0 replied:
Is there a way to execute a pre-defined plan, which we can use that to auto-gen the mix-match the test cases?

walterddr commented on pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/operator/AggregateOperatorTest.java:
```diff
@@ -188,60 +191,27 @@ public void shouldAggregateSingleInputBlockWithLiteralInput() {
-    Mockito.when(merger.init(Mockito.any(), Mockito.any())).thenReturn(1d);
-    DataSchema outSchema = new DataSchema(new String[]{"sum"}, new ColumnDataType[]{DOUBLE});
-    AggregateOperator operator =
-        new AggregateOperator(OperatorTestUtil.getDefaultContext(), _input, outSchema, calls, group, inSchema,
-            ImmutableMap.of("SUM", cdt -> merger));
```
This test is no longer allowed as we do not have a merger API
